packages/create/templates/common/module/control.lua

```diff
@@ -28,7 +30,7 @@ local function bar()
 	game.print("bar")
 end
 
---- Clusterio provides a few custom events, on_server_startup is the most useful and should be used in place of on_load
+--- Clusterio provides a few custom events, on_server_startup is the most useful and should be used in place of on_load and on_init
```

Comment:
This is not a replacement for on_load. Moreover, I think a separate on_load is needed to restore the global alias when joining a multiplayer.. That on_load has to also work on first load (i.e. when the `global["__plugin_name__"] == nil` on the server, or the case where global["__plugin_name__"] still contains premigrated data.

Reply:
> I think a separate on_load is needed to restore the global alias when joining a multiplayer.

As tested [here](https://discord.com/channels/450361298220351489/450361298220351491/1246843044167422042), there are only a few events in which the global alias would be invalid, these events are mentioned in the comment on line 11. It is also safe to setup metatables on global data within this method, as long as the same edge case events are covered. (Although you should use `register_metatable` for this)

> This is not a replacement for on_load.

The only part which is not possible is the registering of conditional event handlers which must still happen within `on_load`. I find this to be an anti pattern within scenario code and should only be used within mods, but I can add this as an additional comment. Any use not mentioned here or the previous comment are considered [illegitimate uses of the event](https://lua-api.factorio.com/latest/auxiliary/data-lifecycle.html#[4]-).

> That on_load has to also work on first load (i.e. when the `global["__plugin_name__"] == nil`)

It is impossible to restore a reference when the value is nil. It is also impossible to modify global to not be nil during on_load. Your only option here is to use on_server_startup and to make sure the edge case events are accounted for. The same is true for "premigrated data" although we currently don't have any method of telling a plugin that it has been updated to a new version, so that is currently undefined behaviour.

Reply:
> As tested here, there are only a few events in which the global alias would be invalid, these events are mentioned in the comment on line 11. It is also safe to setup metatables on global data within this method, as long as the same edge case events are covered. (Although you should use register_metatable for this)

As far as I see it is not called from the on_load() of the module. When joining the game the game is saved, transmitted and loaded by the new player. During the load of the save no method is called. So for the joining player the `globalData` variable will *always* be `{}`. This is a problem as soon as there is actually data stored in the global, as that will not be in `globalData` for the client, but will be for the server. Hence the `on_load()` function [here](https://discord.com/channels/450361298220351489/450361298220351491/1246805038861652080)

> The only part which is not possible is the registering of conditional event handlers which must still happen within on_load. I find this to be an anti pattern within scenario code and should only be used within mods, but I can add this as an additional comment. Any use not mentioned here or the previous comment are considered [illegitimate uses of the event](https://lua-api.factorio.com/latest/auxiliary/data-lifecycle.html#%5B4%5D-).

That link specifies "The [global](https://lua-api.factorio.com/latest/auxiliary/global.html) table can be accessed and is safe to read from, but not write to, as doing so will lead to an error.", yet this method modifies the global table.

> It is impossible to restore a reference when the value is nil. It is also impossible to modify global to not be nil during on_load. Your only option here is to use on_server_startup and to make sure the edge case events are accounted for. The same is true for "premigrated data" although we currently don't have any method of telling a plugin that it has been updated to a new version, so that is currently undefined behaviour.

And that is precisely why I would expect that the on_load() implementation has a giant precautionary warning attached to it.

Reply:
Attempted to make a [module](https://gist.github.com/Cooldude2606/16f48622d3eef1265fc91b39a98faa87) for this and you were correct that on_load still needs to exist, although the handler only needs to do is assign the alias and nothing else, init and migration of global can happen within server startup. I will modify the template to match the patterns used on the working module.